### PR TITLE
Add support for defining custom window frame bounds for window functions

### PR DESCRIPTION
**Pull request by yashmayya**
- Fixes https://github.com/apache/pinot/issues/11406.
- Background reading - https://www.postgresql.org/docs/current/tutorial-window.html, https://www.postgresql.org/docs/current/functions-window.html, https://www.postgresql.org/docs/current/sql-expressions.html#SYNTAX-WINDOW-FUNCTIONS (this one is most relevant to this PR).
- Currently, Pinot's window function implementations have limited or even incorrect support for window frame bounds. For instance, `FIRST_VALUE` / `LAST_VALUE` assume that the window frame is always `ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING` even though the default window frame as per standard SQL is `RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW`. Furthermore, support for defining the lower bound explicitly as `UNBOUNDED PRECEDING` / `CURRENT ROW` / `n FOLLOWING` / `n PRECEDING` and the upper bound as `UNBOUNDED FOLLOWING` / `CURRENT ROW` / `n FOLLOWING` / `n PRECEDING` does not exist.
- This patch adds support for any custom bounds (offset based or otherwise) for `ROWS` window frames, and also adds support for `UNBOUNDED PRECEDING` / `CURRENT ROW` / `UNBOUNDED FOLLOWING` bounds for `RANGE` window frames. There are a ton of edge cases to be handled here but this patch attempts to add test cases to cover most of these scenarios.
- Note that Calcite (and hence Pinot) only supports `ROWS` and `RANGE` based window frame bounds, whereas Postgres also supports `GROUPS`.
- The planner side changes (mainly literal extraction) are built over https://github.com/apache/pinot/pull/14233/.
- This patch also introduces a new `WindowValueAggregator` interface along with implementations for `SUM`, `COUNT`, `MIN`, `MAX`, `BOOLAND`, `BOOLOR`. These use sliding window based aggregation algorithms to efficiently compute aggregations for the aggregate type window functions. This replaces the older `Merger` interface and ensures that the time complexity for computing aggregate window functions for `ROWS` based windows is `O(N)` and not `O(N * K)` (where `N` is the total number of rows and `K` is the window size) which would've been the case if we continued using the `Merger` based approach. We still need to add support for type specific window value aggregators (which was also the case with the mergers) so this isn't a regression.
- Note that all the changes in this patch only affect the aggregate window functions (`SUM`, `COUNT`, `MIN`, `MAX` etc.) and `FIRST_VALUE` / `LAST_VALUE`. The other window functions currently supported by Pinot (`LAG`, `LEAD`, `RANK`, `DENSE_RANK`, `ROW_NUMBER`) don't support custom window frame bounds and Calcite ensures that during query planning.
- Calcite also does some other validation for window frame bounds like making sure lower bound isn't `UNBOUNDED FOLLOWING` / upper bound isn't `UNBOUNDED PRECEDING`, lower bound isn't `n FOLLOWING` if upper bound is `n PRECEDING` and vice versa etc.

### CODE REVIEW

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/window/aggregate/SumWindowValueAggregator.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+package org.apache.pinot.query.runtime.operator.window.aggregate;
+
+import javax.annotation.Nullable;
+
+
+/**
+ * Window value aggregator for SUM window function.
+ */
+public class SumWindowValueAggregator implements WindowValueAggregator<Object> {
+  private Double _sum = 0.0;
```
Comment (Jackie-Jiang) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/window/aggregate/SumWindowValueAggregator.java:28]: Use primitive `double`

**Review by Jackie-Jiang — COMMENTED** [review]
Well done!

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/window/aggregate/AggregateWindowFunction.java [hunk 2]
```diff
@@ -18,113 +18,158 @@
...
-      List<RelFieldCollation> collations, boolean partitionByOnly) {
-    super(aggCall, inputSchema, collations, partitionByOnly);
-    String functionName = aggCall.getFunctionName();
-    Function<ColumnDataType, Merger> mergerCreator = AggregationUtils.Accumulator.MERGERS.get(functionName);
-    Preconditions.checkArgument(mergerCreator != null, "Unsupported aggregate function: %s", functionName);
-    _merger = mergerCreator.apply(_dataType);
+      List<RelFieldCollation> collations, WindowFrame windowFrame) {
+    super(aggCall, inputSchema, collations, windowFrame);
+    _functionName = aggCall.getFunctionName();
+    _windowValueAggregator = WindowValueAggregatorFactory.getWindowValueAggregator(_functionName, _dataType,
```
[the remark this answers is not in the document]
Comment (yashmayya) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/window/aggregate/AggregateWindowFunction.java:43]: Good catch, fixed!

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/window/aggregate/MaxWindowValueAggregator.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import javax.annotation.Nullable;
+
+
+/**
+ * Window value aggregator for MAX window function.
+ */
+public class MaxWindowValueAggregator implements WindowValueAggregator<Object> {
+
+  private final boolean _supportRemoval;
+  private final Deque<Double> _deque = new LinkedList<>();
```
[the remark this answers is not in the document]
Comment (yashmayya) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/window/aggregate/MaxWindowValueAggregator.java]: Makes sense, updated.

File: pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotWindowExchangeNodeInsertRule.java [hunk 4]
```diff
@@ -144,60 +147,88 @@ public void onMatch(RelOptRuleCall call) {
...
       }
     }
-    if (aggCallChanged) {
-      return new Window.Group(oldWindowGroup.keys, oldWindowGroup.isRows, oldWindowGroup.lowerBound,
-          oldWindowGroup.upperBound, oldWindowGroup.orderKeys, newAggCallWindow);
+
+    RexWindowBound lowerBound = oldWindowGroup.lowerBound;
+    RexNode offset = lowerBound.getOffset();
+    if (offset != null) {
+      RexLiteral literal = getLiteral(offset, numInputFields, window.constants, projects);
```
[the remark this answers is not in the document]
Comment (yashmayya) [reply pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotWindowExchangeNodeInsertRule.java:189]: Good point - it should be guaranteed here that offset is a valid literal if offset is non-null (can be null for `UNBOUNDED` and `CURRENT ROW` cases only). For `ROWS` type window frames it has to be an integer literal, and for `RANGE` it can either be a numeric literal or a date / time literal (very few databases currently support this though) like `INTERVAL '1' DAY` - although we currently don't support offset `FOLLOWING / PRECEDING` with `RANGE` type window frames at all so we don't need to worry about this right now. 

I've updated this to throw an exception with a useful error message if we are unable to read the literal for some reason here.

**Review by yashmayya — COMMENTED** [review]
Thanks for the review Jackie! I've made the requested changes.

**Review by Jackie-Jiang — APPROVED** [review]
